src/esp32/pin.rs

```diff
@@ -9,3 +14,143 @@ impl<'d, T: Pin, MODE> PinExt for PinDriver<'d, T, MODE> {
         self.pin()
     }
 }
+
+fn install_gpio_isr_service() -> anyhow::Result<()> {
+    static GPIO_ISR_SERVICE_INSTALLED: Lazy<Arc<OnceCell<()>>> =
+        Lazy::new(|| Arc::new(OnceCell::new()));
+    GPIO_ISR_SERVICE_INSTALLED.get_or_try_init(|| {
+        unsafe {
+            esp!(gpio_install_isr_service(ESP_INTR_FLAG_IRAM as i32))?;
+        };
+        Ok::<(), anyhow::Error>(())
+    })?;
+    Ok(())
+}
+
+/// Esp32GPIOPin is a wrapper for a pin on ESP32 as represented in esp-idf-hal
+/// and esp-idf-sys. This exists so that all micro-RDK drivers can interact
+/// with pins through the board instance and avoid conflicting uses of pins
+/// by multiple processes
+pub struct Esp32GPIOPin {
+    pin: i32,
+    driver: PinDriver<'static, AnyIOPin, InputOutput>,
+    interrupt_type: Option<InterruptType>,
+    event_count: Arc<AtomicU32>,
+    pwm_driver: Option<PwmDriver<'static>>,
+}
+
+impl Esp32GPIOPin {
+    pub fn new(pin: i32, pull: Option<Pull>) -> anyhow::Result<Self> {
+        let mut driver = PinDriver::input_output(unsafe { AnyIOPin::new(pin) })?;
+        if let Some(pull) = pull {
+            driver.set_pull(pull)?;
+        }
+        Ok(Self {
+            pin,
+            driver,
+            interrupt_type: None,
+            event_count: Arc::new(AtomicU32::new(0)),
+            pwm_driver: None,
+        })
+    }
+
+    pub fn pin(&self) -> i32 {
+        self.pin
+    }
+
+    pub fn is_high(&self) -> bool {
+        self.driver.is_high()
+    }
+
+    pub fn set_high(&mut self) -> anyhow::Result<()> {
+        self.driver
+            .set_high()
+            .map_err(|e| anyhow::anyhow!("couldn't set pin {} high {}", self.pin, e))
+    }
+
+    pub fn set_low(&mut self) -> anyhow::Result<()> {
+        self.driver
+            .set_low()
+            .map_err(|e| anyhow::anyhow!("couldn't set pin {} low {}", self.pin, e))
+    }
+
+    pub fn get_pwm_duty(&self) -> f64 {
+        match &self.pwm_driver {
+            Some(pwm_driver) => pwm_driver.get_duty_pct(),
+            None => 0.0,
+        }
+    }
+
+    pub fn set_pwm_duty(&mut self, pct: f64) -> anyhow::Result<()> {
+        match self.pwm_driver.as_mut() {
+            Some(pwm_driver) => pwm_driver.set_duty_pct(pct),
+            None => {
+                let mut pwm_driver = create_pwm_driver(unsafe { AnyIOPin::new(self.pin) }, 10000)?;
+                pwm_driver.set_duty_pct(pct)?;
+                self.pwm_driver = Some(pwm_driver);
+                Ok(())
+            }
+        }
+    }
+
+    pub fn get_pwm_frequency(&self) -> u64 {
+        match &self.pwm_driver {
+            Some(pwm_driver) => pwm_driver.get_frequency() as u64,
+            None => 0,
+        }
+    }
+
+    pub fn set_pwm_frequency(&mut self, freq: u64) -> anyhow::Result<()> {
+        match self.pwm_driver.as_mut() {
```

Comment:
how do we protect against configuring a pin that's already a digital interrupt?

Reply:
ah I simply forgot, I'll just error if `self.interrupt_type` is not `None`